state/event_table_test.go

```diff
@@ -946,3 +979,110 @@ func TestEventTableSelectUnknownEventIDs(t *testing.T) {
 		}
 	}
 }
+
+func TestEventTableUpdateBeforeSnapshotIDs(t *testing.T) {
```

Comment:
You're going to need a lot more tests to convince me that the SQL is fine. Specifically:
 - `UpdateBeforeSnapshotIDs` with no `beforeSnapshotUpdate`
 - `UpdateBeforeSnapshotIDs` with exactly 1 `beforeSnapshotUpdate`
 - `UpdateBeforeSnapshotIDs` with `len(beforeSnapshotUpdate)` > chunk size.
 -  `beforeSnapshotUpdate` has some known and some unknown event NIDs.
 - `beforeSnapshotUpdate` has some duplicate event NIDs and some not duplicated.

Reply:
That's very fair. I am also suspicious of this arcane incantation

Reply:
> beforeSnapshotUpdate has some duplicate event NIDs and some not duplicated.

I think this is essentially undefined behaviour. Quoting postgres' [docs](https://www.postgresql.org/docs/current/sql-update.html#id-1.9.3.183.8):

> When a FROM clause is present, what essentially happens is that the target table is joined to the tables mentioned in the from_item list, and each output row of the join represents an update operation for the target table. When using FROM you should ensure that the join produces at most one output row for each row to be modified. In other words, a target row shouldn't join to more than one row from the other table(s). If it does, then only one of the join rows will be used to update the target row, but which one will be used is not readily predictable.